File: src/DeviceIoZmx.cpp
// this file is covered by the  GNU LESSER GENERAL PUBLIC LICENSE Version 3 or later
// please see LICENSE.txt for more details and licensing issues
// copyright Etienne de Foras ( the author )  mailto: etienne.deforas@gmail.com

#include <sstream>
#include <fstream>
#include <string>
#include <cassert>

using namespace std;

#include "DeviceIoZmx.h"
#include "OpticalDevice.h"

/////////////////////////////////////////////////////////////////////////////////////////////////
// utf16/utf8 utilies ( partially from stackoverflow)

///////////////////////////////////////////////////////////////////////////////////////////
class TextFileReader //handle utf8 or utf16 encoding
{
public:
	TextFileReader()
	{
		_bIsUtf16=false;
	}

	bool open(string sFile)
	{
		_bIsUtf16 = is_utf16(sFile);

		if(_bIsUtf16)
		{
			wistr.open(sFile,ios::binary);

			wchar_t bom;
			wistr >> bom;

			return !(!wistr);
		}
		else
		{
			istr.open(sFile);
			return !(!istr);
		}
	}
	
	bool eof()
	{
		if (_bIsUtf16)
			return wistr.eof();
		else
			return istr.eof();
	}

	bool getline(string& s)
	{
		bool bOk;
		if (_bIsUtf16)
		{
			wstring ws;
			std::getline(wistr, ws);
			s = to_string(ws);

			bOk= !(!wistr);
		}
		else
		{
			std::getline(istr, s);
			bOk= !(!istr);
		}

		size_t iPosStart = s.find_first_not_of(" \n\r\t");
		size_t iPosEnd = s.find_last_not_of(" \n\r\t");
		if (iPosStart != string::npos)
			s = s.substr(iPosStart, iPosEnd - iPosStart + 1);

		return bOk;
	}

	bool is_utf16(string sFile)
	{
		ifstream ifs(sFile, ios::binary);
		while (!ifs.eof())
		{
			unsigned char c;
			ifs >> c;
			if (c == 0)
				return true;
		}
		return false;
	}

	std::string to_string(const std::wstring& wstr)
	{
		//raw way to convert from wstring...
		string str;
		for (size_t i = 0; i < wstr.size(); i++)
		{
			wchar_t w = wstr[i];
			if ( (w<127) && (w>0) )
				str.push_back((char)w);
		}

		return str;
	}

private:
	bool _bIsUtf16;
	ifstream istr;
	wifstream wistr;
};

////////////////////////////////////////////////////////////////////////////////
OpticalDevice* DeviceIoZmx::import(const string& sFile)
{
	TextFileReader tfr;
	if (!tfr.open(sFile))
		return nullptr;

    OpticalDevice* pOD=new OpticalDevice();
    pOD->set_relative_convention(true);

	bool bMustUseColoredLight = false;
    double dDimensionFactor=1.;
    int iSurface=0;
    bool bSurfPending=false;
    string sType;
    double dCurvature=0.;
    double dTick=0.;
    double dDiameter=0.;
    double dConic=0.;
    double dInnerDiameter=0.;
    double dApertureDiameter=-1.;
	double dHalfFOV=0;
    string sComment;
    string sNote="Imported from ZMX file: "+ sFile+"\nWarning, import may be inaccurate!\n\n";
    string sGlassCatalog;

    while(!tfr.eof())
    {
		string sLine;
		tfr.getline(sLine);

        //trim first space if any
        size_t iPos=sLine.find_first_not_of(" ");
        if(iPos!=string::npos)
            sLine=sLine.substr(iPos);

        if(sLine.size()<4)
            continue; // nothing to parse

        string sKey=sLine.substr(0,4);

        string sVal=sLine.substr(4); // can be empty

        //trim first space if any
        size_t iPosStart=sVal.find_first_not_of(" \n\r");
		if(iPosStart !=string::npos)
            sVal=sVal.substr(iPosStart);
			
        if((sKey=="SURF") || (sKey=="BLNK"))
        {
            if(bSurfPending)
            {
                pOD->insert_surface(iSurface);
                pOD->set_comment(iSurface,sComment);
                pOD->set_type(iSurface,sType);
                pOD->set(iSurface,THICK,dDimensionFactor*dTick);
                pOD->set(iSurface,CURVATURE,dCurvature/dDimensionFactor);
                pOD->set(iSurface,DIAMETER,dDiameter*dDimensionFactor);
                pOD->set(iSurface,CONIC,dConic);
                pOD->set(iSurface,INNER_DIAMETER,dInnerDiameter*dDimensionFactor);
                bSurfPending=false;
            }

            if(sKey=="SURF") //new surf
            {
                iSurface=-1;
                stringstream ss(sVal); ss >> iSurface;

                if(iSurface==-1)
                { delete pOD; return 0; } // not a valid field

                bSurfPending=true;

                dCurvature=0.;
                dDiameter=0.;
                sType="Air";
                dConic=0.;
                sComment="";
                dInnerDiameter=0.;
            }
        }

		if (sKey == "YFLD") //input light field of view (in degree for now)
		{
			stringstream ss(sVal);
			while (!ss.eof())
			{
				double d = 0.;
				ss >> d;
				if (d > dHalfFOV)
					dHalfFOV = d;
			}
		}

        if(sKey=="UNIT")
        {
            string sFirstVal=sVal.substr(0,sVal.find_first_of(" "));

            if(sFirstVal=="MM")
                dDimensionFactor=1.;

            if(sFirstVal=="METER")
                dDimensionFactor=1.e3;

            if(sFirstVal=="CM")
                dDimensionFactor=10.;

            if(sFirstVal=="IN")
                dDimensionFactor=25.4;
        }

        if(sKey=="TYPE")
        {
            if( (sVal!="STANDARD") && (sVal!="EVENASPH") )
            { delete pOD; return 0; } // dont' know how to parse
        }

        if(sKey=="CURV") //rcurv
        {
            stringstream ss(sVal); ss >> dCurvature;
        }

        if(sKey=="COMM") //comment
        {
            sComment=sVal;
        }

		if (sKey == "STOP") //for example, Aperture stop
		{
			sType="stop";
		}
		
		
		if(sKey=="NOTE") //note
        {
            string sNoteVal;
            auto iPosSpace=sVal.find_first_of(" ");
            if(iPosSpace!=string::npos)
                sNoteVal=sVal.substr(iPosSpace,string::npos);

            sNote+=sNoteVal+"\n";
        }

        if(sKey=="DISZ") //thicks
        {
			if (sVal == "INFINITY")
				dTick = 0.;
			else
			{
				stringstream ss(sVal);
				ss >> dTick;
			}
		}

        if(sKey=="GCAT")
        {
            stringstream ss(sVal); ss >> sGlassCatalog; //TODO use
        }

        if(sKey=="OBSC") //inner diameter
        {
            double dTemp;
            stringstream ss(sVal); ss >> dTemp >> dInnerDiameter;
        }

        if(sKey=="CONI") //conic
        {
            stringstream ss(sVal); ss >> dConic;
        }

        if(sKey=="DIAM") //diameter
        {
            stringstream ss(sVal); ss >> dDiameter;
            dDiameter*=2.;
        }

        if(sKey=="ENPD") //aperture diameter
        {
            stringstream ss(sVal); ss >> dApertureDiameter;
        }

        if(sKey=="GLAS") //type or glass
        {
            string sFirstWord;
            auto iPosSpace=sVal.find_first_of(" ");
            if(iPosSpace!=string::npos)
                sFirstWord=sVal.substr(0,iPosSpace);

            if(sFirstWord=="MIRROR")
                sType="reflect";
			else
			{
				sType = sFirstWord; //TODO more robust test
				bMustUseColoredLight=true;
			}
		}
    }

	if (bSurfPending)
	{
		pOD->insert_surface(iSurface);
		pOD->set_comment(iSurface, sComment);
		pOD->set_type(iSurface, sType);
		pOD->set(iSurface, THICK, dDimensionFactor*dTick);
		pOD->set(iSurface, CURVATURE, dCurvature / dDimensionFactor);
		pOD->set(iSurface, DIAMETER, dDiameter*dDimensionFactor);
		pOD->set(iSurface, CONIC, dConic);
		pOD->set(iSurface, INNER_DIAMETER, dInnerDiameter*dDimensionFactor);
		bSurfPending = false;
	}
	assert(!bSurfPending);

    if(!sNote.empty())
        pOD->set_note(sNote);

    // set the last surface type as image, TODO better check
	int iLastSurface = pOD->nb_surface() - 1;
	if(iLastSurface >=0)
    {    
		if (pOD->type(iLastSurface) == "Air")
		{
			pOD->set_type(iLastSurface, "image");
			pOD->set(iLastSurface, AUTO_DIAMETER, 1.);
			pOD->set_image_autocurvature(true);
		}
        if(dApertureDiameter>=0.)
            pOD->set(0,DIAMETER,dApertureDiameter*dDimensionFactor);
    }

	// add colored light in case of glass, only visible light for now
	if (bMustUseColoredLight)
		pOD->set_light_colors("Red.Yellow.Green.Blue.");

	//field of view
	pOD->set_half_field_of_view(dHalfFOV);

    return pOD;
}
////////////////////////////////////////////////////////////////////////////////

File: python/DeviceIOZmx.py
import io
import os
import OpticalDevice

class TextFileReader:
    """Handle utf8 or utf16 encoding"""
    def __init__(self):
        self._is_utf16 = False
        self._file = None
    
    def open(self, file_path):
        self._is_utf16 = self._is_utf16_file(file_path)
        
        if self._is_utf16:
            self._file = open(file_path, 'rb')
            # Read and skip BOM if present
            bom = self._file.read(2)
            return True
        else:
            self._file = open(file_path, 'r', encoding='utf-8')
            return True
    
    def eof(self):
        pos = self._file.tell()
        is_eof = not self._file.read(1)
        self._file.seek(pos)
        return is_eof
    
    def getline(self):
        line = ''
        if self._is_utf16:
            # Simplified UTF-16 handling - this would need proper implementation
            # for real UTF-16 files
            line = self._file.readline().decode('utf-16le', errors='ignore')
        else:
            line = self._file.readline()
        
        if line:
            line = line.strip()
        return line
    
    def _is_utf16_file(self, file_path):
        with open(file_path, 'rb') as f:
            while True:
                chunk = f.read(1024)
                if not chunk:
                    break
                if b'\x00' in chunk:
                    return True
        return False
    
    def close(self):
        if self._file:
            self._file.close()

class DeviceIoZmx:
    @staticmethod
    def import_zmx(file_path):
        tfr = TextFileReader()
        if not tfr.open(file_path):
            return None
        
        od = OpticalDevice.OpticalDevice()
        od.set_relative_convention(True)
        
        must_use_colored_light = False
        dimension_factor = 1.0
        current_surface = 0
        surface_pending = False
        surface_type = ""
        curvature = 0.0
        thickness = 0.0
        diameter = 0.0
        conic = 0.0
        inner_diameter = 0.0
        aperture_diameter = -1.0
        half_fov = 0.0
        comment = ""
        note = f"Imported from ZMX file: {file_path}\nWarning, import may be inaccurate!\n\n"
        glass_catalog = ""
        
        while not tfr.eof():
            line = tfr.getline()
            if not line or len(line) < 4:
                continue
            
            key = line[:4].upper()
            value = line[4:].strip()
            
            if key in ("SURF", "BLNK"):
                if surface_pending:
                    od.insert_surface(current_surface)
                    od.set_comment(current_surface, comment)
                    od.set_type(current_surface, surface_type)
                    od.set(current_surface, "THICK", dimension_factor * thickness)
                    od.set(current_surface, "CURVATURE", curvature / dimension_factor)
                    od.set(current_surface, "DIAMETER", diameter * dimension_factor)
                    od.set(current_surface, "CONIC", conic)
                    od.set(current_surface, "INNER_DIAMETER", inner_diameter * dimension_factor)
                    surface_pending = False
                
                if key == "SURF":
                    try:
                        current_surface = int(value)
                    except ValueError:
                        tfr.close()
                        return None
                    
                    if current_surface == -1:
                        tfr.close()
                        return None
                    
                    surface_pending = True
                    curvature = 0.0
                    diameter = 0.0
                    surface_type = "Air"
                    conic = 0.0
                    comment = ""
                    inner_diameter = 0.0
            
            elif key == "YFLD":
                try:
                    values = [float(v) for v in value.split()]
                    if values:
                        current_fov = max(values)
                        if current_fov > half_fov:
                            half_fov = current_fov
                except ValueError:
                    pass
            
            elif key == "UNIT":
                unit = value.split()[0].upper() if value else ""
                if unit == "MM":
                    dimension_factor = 1.0
                elif unit == "METER":
                    dimension_factor = 1e3
                elif unit == "CM":
                    dimension_factor = 10.0
                elif unit == "IN":
                    dimension_factor = 25.4
            
            elif key == "TYPE":
                if value not in ("STANDARD", "EVENASPH"):
                    tfr.close()
                    return None
            
            elif key == "CURV":
                try:
                    curvature = float(value)
                except ValueError:
                    pass
            
            elif key == "COMM":
                comment = value
            
            elif key == "STOP":
                surface_type = "stop"
            
            elif key == "NOTE":
                note_value = value[value.find(" ") + 1:] if " " in value else value
                note += note_value + "\n"
            
            elif key == "DISZ":
                if value.upper() == "INFINITY":
                    thickness = 0.0
                else:
                    try:
                        thickness = float(value)
                    except ValueError:
                        pass
            
            elif key == "GCAT":
                glass_catalog = value
            
            elif key == "OBSC":
                try:
                    parts = value.split()
                    if len(parts) >= 2:
                        inner_diameter = float(parts[1])
                except ValueError:
                    pass
            
            elif key == "CONI":
                try:
                    conic = float(value)
                except ValueError:
                    pass
            
            elif key == "DIAM":
                try:
                    diameter = float(value) * 2.0
                except ValueError:
                    pass
            
            elif key == "ENPD":
                try:
                    aperture_diameter = float(value)
                except ValueError:
                    pass
            
            elif key == "GLAS":
                first_word = value.split()[0] if value else ""
                if first_word == "MIRROR":
                    surface_type = "reflect"
                else:
                    surface_type = first_word
                    must_use_colored_light = True
        
        tfr.close()
        
        if surface_pending:
            od.insert_surface(current_surface)
            od.set_comment(current_surface, comment)
            od.set_type(current_surface, surface_type)
            od.set(current_surface, "THICK", dimension_factor * thickness)
            od.set(current_surface, "CURVATURE", curvature / dimension_factor)
            od.set(current_surface, "DIAMETER", diameter * dimension_factor)
            od.set(current_surface, "CONIC", conic)
            od.set(current_surface, "INNER_DIAMETER", inner_diameter * dimension_factor)
            surface_pending = False
        
        if note:
            od.set_note(note)
        
        # Set the last surface type as image
        last_surface = od.nb_surface() - 1
        if last_surface >= 0:
            if od.type(last_surface) == "Air":
                od.set_type(last_surface, "image")
                od.set(last_surface, "AUTO_DIAMETER", 1.0)
                od.set_image_autocurvature(True)
            
            if aperture_diameter >= 0:
                od.set(0, "DIAMETER", aperture_diameter * dimension_factor)
        
        # Add colored light in case of glass
        if must_use_colored_light:
            od.set_light_colors("Red.Yellow.Green.Blue.")
        
        # Set field of view
        od.set_half_field_of_view(half_fov)
        
        return od
